### PR TITLE
Fix DDP batch size handling

### PR DESCRIPTION
**Pull request by buddit1**
Builds on #352 to finish implementing DDP. Follow up work will address performance improvements, priority here is correct implementation.

### Problem Being Addressed
Previously the number of steps per epoch and total number of positions seen per epoch were dependent on the number of GPUs even when keeping all other settings the same. This made training results differ significantly when keeping all settings fixed but changing the number of GPUs. 

### Changes Introduced Between #352 and this PR

1. `--batch-size` now specifies the global batch size i.e. number of samples used per optimizer step. 
    This automatically gets converted into the per gpu batch size. e.g. for `--batch-size=65536` 
  * 1 GPU -> 65536 per GPU per batch
  * 2 GPU -> 32768 per GPU per batch
  * 4 GPU -> 16384 per GPU per batch
  If `--batch-size` can't be evenly split between the GPUs a `ValueError` is raised.
2. `--gpus` default behavior changed. 
Previously the default behavior was to set devices to "auto". 
New default behavior is to only train on device 0.
This change is to remove any ambiguity in the batch size handling. I'd like to reintroduce "auto" in future work but reliably handling the batch size conversion with "auto" is a bit more complex and warrants testing.
4. Updated timing of calling `SparseBatchDataset.__iter__` 
from `FixedNumBatchesDataset.__init__` to  `FixedNumBatchesDataset._start_prefetching`. 
this means that DDP is reliably initialized at this stage. As a result the env var work arounds for accessing rank and world size was removed.

## Testing

I verified batch size behavior on 1, 2, and 4 GPUs. 
I also ran 20 epochs of training on 1 and 4 a100s connected by NVLINK.

Logs are included here:
[ddp_experiment_logs_no_checkpoints.zip](https://github.com/user-attachments/files/25461751/ddp_experiment_logs_no_checkpoints.zip)

Directory Structure:
```
verify_correctness/
├── 1_gpu.log
├── 4_gpu.log
├── check_batch_size_handling
│   ├── 1_gpu.log
│   ├── 2_gpu.log
│   ├── 4_gpu.log
│   └── notes.md
├── lightning_logs
│   ├── version_0
│   │   ├── checkpoints
│   │   ├── events.out.tfevents.1771704615.2c4bc1093d12.32802.0
│   │   └── hparams.yaml
│   └── version_1
│       ├── checkpoints
│       ├── events.out.tfevents.1771706174.2c4bc1093d12.34263.0
│       └── hparams.yaml
└── training_finished
```

Each .log file includes the full `train.py` command at the top. All of the .log files were created by modifying the script [log_train.sh](https://github.com/buddit1/nnue-pytorch/blob/verify-batchsize-ddp-2026-02-21/scripts/log_train.sh) 

### Verifying Batch Size Handling
The files in the subdirectory `verify_correctness/check_batch_size_handling` were created using the branch [buddit1/verify-batchsize-ddp-2026-02-21](https://github.com/buddit1/nnue-pytorch/tree/verify-batchsize-ddp-2026-02-21) they include print output at the bottom of each log file showing the per gpu batch size. (see [model/lightning_module.py](https://github.com/buddit1/nnue-pytorch/blob/verify-batchsize-ddp-2026-02-21/model/lightning_module.py#L118) for the exact code changes used to verify the batch size)

### Training Consistency Check
Ran 20 epochs with:
* 1 GPU -> version_0
* 4 GPUs -> version_1
* Hardware was 4xA100s connected with NVLINK

Full args used are in
* verify_correctness/1_gpu.log
* verify_correctness/4_gpu.log

Results:
* steps per epoch are now independent of GPU count.
* Loss curves look to be consistent within noise
* Wall clock times are included in the .log files but these aren't intended as a meaningful benchmark.

### CODE REVIEW

**Comment by vondele** [comment]
Also on the more realistic setup, loss seems to be  correct:
```
numactl --cpunodebind=0 python -u train.py /workspace/data/vondele/from_kaggle_2/T60T70wIsRightFarseerT60T74T75T76.split_0.binpack /workspace/data/vondele/from_kaggle_2/T60T70wIsRightFarseerT60T74T75T76.split_1.binpack /workspace/data/vondele/from_kaggle_2/T60T70wIsRightFarseerT60T74T75T76.split_2.binpack /workspace/data/vondele/from_kaggle_2/T60T70wIsRightFarseerT60T74T75T76.split_3.binpack /workspace/data/vondele/from_kaggle_2/T60T70wIsRightFarseerT60T74T75T76.split_4.binpack /workspace/data/official-stockfish/master-binpacks/nodes5000pv2_UHO.binpack /workspace/data/official-stockfish/master-binpacks/wrongIsRight_nodes5000pv2.binpack /workspace/data/official-stockfish/master-binpacks/multinet_pv-2_diff-100_nodes-5000.binpack /workspace/data/official-stockfish/master-binpacks/dfrc_n5000.binpack --threads=4 --num-workers=96 --batch-size=65536 --features=Full_Threats^ --l1=1024 --lr=4.375e-4 --gamma=0.995 --start-lambda=1.0 --end-lambda=0.75 --random-fen-skipping=10 --early-fen-skipping=12 --pc-y1=0.6893201149773951 --pc-y2=2.9285769485515805 --pc-y3=1.4386005301749225 --w1=3.3553547771220007 --w2=0.7006821612968052 --max_time=00:11:30:00 --default_root_dir /workspace/scratch/training/runs/run_2  --max_epochs=8 --gpus=0,1,2,3
```
Speedup is visible but small
39.01 it/s -> 47.6 it/s
but as said this can be looked into in a later iteration.

In this case GPU util is around 60%.

File: .gitignore [hunk 1]
```diff
@@ -5,4 +5,8 @@ logs/
 testing
 libtraining_data_loader.*
 easy_train_data
-*.yaml
\ No newline at end of file
+*.yaml
+CMakeCache.txt
```
Comment (TonyCongqianWang) [inline .gitignore]: Not sure if the `.gitignore` changes were intended

Comment (buddit1) [reply .gitignore]: I believe the gitignore changes all came from @Disservin's work on #352. I'm not sure if they were intended to merge or not but I'd assumed yes.

Comment (TonyCongqianWang) [reply .gitignore]: `CMakeCache.txt` should basically never appear in a projects root I think. Maybe this is an indication for some misconfigured build directory?

Comment (buddit1) [reply .gitignore]: Seems very possible it was just leftover from early experimental work

File: train.py [hunk 2]
```diff
@@ -493,13 +507,16 @@ def main():
         save_top_k=-1,
     )
 
+    # PL hack, undo slurm cluster detection which is broken for us. 'force interactive mode'
+    # see lightning/fabric/plugins/environments/slurm.py near line 110
+    os.environ["SLURM_JOB_NAME"] = "bash"
```
[the remark this answers is not in the document]
Comment (vondele) [reply train.py:380]: fine as is.

File: .gitignore (hunk 1 above)
Comment (buddit1) [reply .gitignore]: I could verify it's not important and remove it in a bit or you/anyone else is certainly welcome to clean it up.

Comment (Disservin) [reply .gitignore]: just get rid of it, dunno what I did back then
